addons/website/models/website_menu.py

```diff
@@ -0,0 +1,157 @@
+
+
+class Menu(models.Model):
+
+    _name = "website.menu"
+    _description = "Website Menu"
+
+    _parent_store = True
+    _order = "sequence, id"
+
+    def _default_sequence(self):
+        menu = self.search([], limit=1, order="sequence DESC")
+        return menu.sequence or 0
+
+    name = fields.Char('Menu', required=True, translate=True)
+    url = fields.Char('Url', default='')
+    page_id = fields.Many2one('website.page', 'Related Page', ondelete='cascade')
+    new_window = fields.Boolean('New Window')
+    sequence = fields.Integer(default=_default_sequence)
+    website_id = fields.Many2one('website', 'Website')
+    parent_id = fields.Many2one('website.menu', 'Parent Menu', index=True, ondelete="cascade")
+    child_id = fields.One2many('website.menu', 'parent_id', string='Child Menus')
+    parent_path = fields.Char(index=True)
+    is_visible = fields.Boolean(compute='_compute_visible', string='Is Visible')
+
+    @api.multi
+    def name_get(self):
+        res = []
+        for menu in self:
+            website_suffix = '%s - %s' % (menu.name, menu.website_id.name)
+            res.append((menu.id, website_suffix if menu.website_id and self.env.user.has_group('website.group_multi_website') else menu.name))
+        return res
+
+    @api.model
+    def create(self, vals):
+        ''' In case a menu without a website_id is trying to be created, we duplicate
+            it for every website.
+            Note: Particulary useful when installing a module that adds a menu like
+                  /shop. So every website has the shop menu.
+                  Be careful to return correct record for ir.model.data xml_id in case
+                  of default main menus creation.
+        '''
+        # Only used when creating website_data.xml default menu
+        if vals.get('url') == '/default-main-menu':
+            return super(Menu, self).create(vals)
+
+        if 'website_id' in vals:
+            return super(Menu, self).create(vals)
+        elif self._context.get('website_id'):
+            vals['website_id'] = self._context.get('website_id')
+            return super(Menu, self).create(vals)
+        else:
+            # create for every site
+            for website in self.env['website'].search([]):
+                w_vals = dict(vals, **{
+                    'website_id': website.id,
+                    'parent_id': website.menu_id.id,
+                })
+                res = super(Menu, self).create(w_vals)
+            # if creating a default menu, we should also save it as such
+            default_menu = self.env.ref('website.main_menu', raise_if_not_found=False)
+            if default_menu and vals.get('parent_id') == default_menu.id:
+                res = super(Menu, self).create(vals)
+        return res  # Only one record is returned but multiple could have been created
+
+    @api.multi
+    def unlink(self):
+        default_menu = self.env.ref('website.main_menu', raise_if_not_found=False)
```

Comment:
```suggestion
        default_menu = self.env.ref('website.main_menu', raise_if_not_found=False)
        menus_to_remove = self
        for menu in self.filtered(lambda m: default_menu and m.parent_id.id == default_menu.id):
            self.env['website.menu'].search([('url', '=', menu.url), ('id', '!=', menu.id)]).unlink()
        return super(Menu, self).unlink()
            menus_to_remove |= self.env['website.menu'].search([('url', '=', menu.url),
                                                                ('website_id', '!=', False),
                                                                ('id', '!=', menu.id)])
        return super(Menu, menus_to_remove).unlink()
```

as for https://github.com/odoo/odoo/commit/8c5ec851fcc42887d5cf4f2d17ba29136f33f5b7 not ported yet to `master`